snap/implicit.go

```diff
@@ -79,3 +85,54 @@ func makeImplicitSlot(snapInfo *Info, ifaceName string) *SlotInfo {
 		Interface: ifaceName,
 	}
 }
+
+// addImplicitHooks adds hooks from the installed snap's hookdir to the snap info.
+//
+// Existing hooks (i.e. ones defined in the YAML) are not changed; only missing
+// hooks are added.
+func addImplicitHooks(snapInfo *Info) error {
+	// First of all, check to ensure the hooks directory exists. If it doesn't,
+	// it's not an error-- there's just nothing to do.
+	hooksDir := snapInfo.HooksDir()
+	if _, err := os.Stat(hooksDir); os.IsNotExist(err) {
+		return nil
+	}
+
+	fileInfos, err := ioutil.ReadDir(hooksDir)
+	if err != nil {
+		return fmt.Errorf("unable to read hooks directory: %s", err)
+	}
+
+	for _, fileInfo := range fileInfos {
+		addHookName(snapInfo, fileInfo.Name())
+	}
+
+	return nil
+}
+
+// addImplicitHooksFromContainer adds hooks from the snap file's hookdir to the snap info.
+//
+// Existing hooks (i.e. ones defined in the YAML) are not changed; only missing
+// hooks are added.
+func addImplicitHooksFromContainer(snapInfo *Info, snapf Container) error {
```

Comment:
Pardon my ignorance here, but while reading this, I was wondering when we are in a situation where we need to read the hooks from a container instead of from a mounted directory. We do the `mount` step in the install/upgrade task pretty early, so I'm curious how /whenhooks will be used before that step?